File: NZOI/2024-2-5-subtask2.cpp
#include <iostream>
#include <utility>
#include <vector>

using namespace std;

const int MAXN = 1005; // Assuming N <= 1000 for Subtask 2

int N, M, K, U;
vector<pair<int, int>> parent(MAXN, {-1, 0});
vector<int> trout_count(MAXN, 0);

void add_trout(int position, int distance) {
  while (distance >= 0) {
    trout_count[position]++;
    if (position == 0)
      break;
    distance -= parent[position].second;
    position = parent[position].first;
  }
}

int main() {
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);

  cin >> N >> M >> K >> U;

  for (int i = 0; i < N - 1; i++) {
    int a, b, w;
    cin >> a >> b >> w;
    parent[b] = {a, w};
  }

  for (int i = 0; i < M; i++) {
    int L, D;
    cin >> L >> D;
    add_trout(L, D);
  }

  for (int i = 0; i < K; i++) {
    char query_type;
    cin >> query_type;
    if (query_type == 'Q') {
      int q;
      cin >> q;
      cout << trout_count[q] << '\n';
    } else {
      int L, D;
      cin >> L >> D;
      add_trout(L, D);
    }
  }

  return 0;
}

File: NZOI/2024-2-5-subtask2.py
N, M, K, U = map(int, input().split())
parent = [-1] * N
for i in range(N - 1):
    a, b, w = map(int, input().split())
    parent[b] = (a, w)
trout_count = [0] * N
def add_trout(position, distance):
    while distance >= 0:
        trout_count[position] += 1
        if position == 0:
            break
        distance -= parent[position][1]
        position = parent[position][0]
for i in range(M):
    L, D = map(int, input().split())
    add_trout(L, D)
for i in range(K):
    query = input().split()
    if query[0] == 'Q':
        print(trout_count[int(query[1])])
    else:
        L, D = map(int, query[1:])
        add_trout(L, D)
